### PR TITLE
[GlobalOptimization] Support SetEncoding on batch matmul cases with p…

### PR DESCRIPTION
**Pull request by Max191**
…roducer CastOpInterface ops

This commit allows SetEncoding to match on CastOpInterface ops that are producers of BatchMatmulOps. This allows inferring the correct input types when the input casting is not implicit inside the BatchMatmul body. This also gives a way to infer signedness on input types through the specific CastOpInterface op types.

### CODE REVIEW

File: compiler/src/iree/compiler/GlobalOptimization/Utils.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/Utils.cpp]: The function doc should only exist in header file. The comment in `cpp` file is more for implementation details. Please remove the function doc from `.cpp` file. Same for other functions.

File: compiler/src/iree/compiler/GlobalOptimization/Utils.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/Utils.cpp]: Can we remove this helper? It is only used here. We can directly write `IntegerType::get(ctx, bitWidth, IntegerType::SignednessSemantics::Unsigned)`. What do you think?

File: compiler/src/iree/compiler/GlobalOptimization/Utils.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/Utils.cpp]: 2023

File: compiler/src/iree/compiler/GlobalOptimization/Utils.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/Utils.cpp]: I think what you're looking for is `getElementTypeOrSelf(castOp->getOperand(0).getType());`

File: compiler/src/iree/compiler/GlobalOptimization/Utils.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/Utils.cpp]: I think you can merge them into a single line: `auto castOp = input.getDefiningOp<CastOpInterface>();`

File: compiler/src/iree/compiler/GlobalOptimization/Utils.h
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/Utils.h]: Lets change the function signature to `FailureOr<CastOpInterface>` or `std::optional<CastOpInterface>`. Though the interface itself does not provide useful utils, but it makes types/readability more clear.

File: compiler/src/iree/compiler/GlobalOptimization/SetEncoding.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/SetEncoding.cpp]: This can be simplified to below:

```
operandTypes[0] = cast<RankedTensorType>(operandTypes[0]).clone(lhsElemType);
```

If it is from cast op, the element type is the result of `getCastElemType()`. If it is not from a cast op, you'll use the old element type. This matches the value of `lhs/rhsElemType`.

File: compiler/src/iree/compiler/GlobalOptimization/Utils.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/Utils.cpp]: We prefer early-return to save nested levels. I.e.,

```cpp
auto tensorType = ...
if (!tensorType) return {};
...
```

File: compiler/src/iree/compiler/GlobalOptimization/SetEncoding.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/SetEncoding.cpp]: I think it is better to capture `std::optional<CastOpInterface>` here, so you don't need to get the defining op again in other functions. So this can be 

```cpp
std::optional<CastOpInterface> maybeLhsCastOp = getDefiningCastOp(origLhs);
Type lhsElemType = maybeLhsCastOp ? getCastElemType(maybeLhsCastOp.value()) : getElemType(origLhs);
```

Then you can pass the optional CastOpInterface to `padAndSetEncoding`

File: compiler/src/iree/compiler/GlobalOptimization/SetEncoding.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/SetEncoding.cpp]: If we replace `bool castResult` with `std::optional<CastOpInterface> srcCastOp`, we can simplify it to `Value padSource = srcCastOp ? srcCastOp->getOperand(0) ? source;`

File: compiler/src/iree/compiler/GlobalOptimization/SetEncoding.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/SetEncoding.cpp]: If `std::optional<CastOpInterface> srcCastOp` is passed to the function, you don't need to call `getDefiningCastOp` again.

**Review by hanhanW — CHANGES_REQUESTED** [review]

**Comment by bjacob** [comment]
I'll wait for a revision of this PR addressing @hanhanW 's comment, but I just wanted to say that I've read the `set_encoding.mlir` testcase and yes this is exactly the result that we want to achieve here.

**Comment by hanhanW** [comment]
> I'll wait for a revision of this PR addressing @hanhanW 's comment, but I just wanted to say that I've read the `set_encoding.mlir` testcase and yes this is exactly the result that we want to achieve here.

Good to know! Yes, the lit test looks really good to me.

**Review by bjacob — APPROVED** [review]

File: compiler/src/iree/compiler/GlobalOptimization/SetEncoding.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/SetEncoding.cpp]: do we need `castResult` being passed? Can we get it from `getElementTypeOrSelf(castOp->getResultTypes()[0])`?

**Review by hanhanW — APPROVED** [review]
Nice work, thanks for addressing all the comment!

**Comment by Max191** [comment]
Let's wait to merge this. I discovered that the encoded `CastOpInterface` ops do not get materialized in MaterializeEncoding. I'm working on that now, and I'll put out another PR once it's ready. We can wait until the MaterializeEncoding changes are done to merge this

**Comment by hanhanW** [comment]
> Let's wait to merge this. I discovered that the encoded `CastOpInterface` ops do not get materialized in MaterializeEncoding. I'm working on that now, and I'll put out another PR once it's ready. We can wait until the MaterializeEncoding changes are done to merge this

Can you add the pattern to https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/Codegen/Common/MaterializeEncodingIntoPackUnPack.cpp and add tests to https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/Codegen/Common/CPU/test/llvmcpu_materialize_encoding.mlir ? I'm going to move patterns from LinalgExt to Codegen/Common/ and I hope that won't impact your work

File: compiler/src/iree/compiler/GlobalOptimization/SetEncoding.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/GlobalOptimization/SetEncoding.cpp:146]: Please add some comments here.

**Review by MaheshRavishankar — COMMENTED** [review]
This is a very specific path being plumbed through... Its OK for now, but we'll very soon hit limitations of this. I am not suggesting alternatives yet cause I dont have a full working path for something more general, so this is OK for now, but just laying a marker down saying this is trying to walk a very fine line.

**Comment by Max191** [comment]
Rebased on main and resolved some conflicts. @bjacob the conflict was involving your recent PR https://github.com/openxla/iree/pull/15431, and I don't see any tests in SetEncoding for it. It was a pretty simple conflict, so I assume it is okay, but just wanted to confirm that it looks good before merging.

**Comment by bjacob** [comment]
Thanks for checking. This looks good!

**Pull request merged**
